modules/tools/rosbag/body_sensation_evaluation.py

```diff
@@ -0,0 +1,224 @@
+            mode["end_time"] = 0.0;  
+            for topic, msg, _t in bag.read_messages([kChassisTopic]):
+                t = long(str(_t))*pow(10,-9)
+                cur_status = msg.driving_mode
+                if mode["status"] != cur_status:
+                    if mode["status"] != 'UNKNOW':
+                        self.driving_mode.append(mode)
+                    mode["status"] = cur_status
+                    mode["start_time"] = t
+                mode["end_time"] = t
+            self.driving_mode.append(mode)
+
+    def _check_status(self, timestamp):
+        """check driving mode according to timestamp"""
+        for mode in self.driving_mode:
+            if timestamp >= mode["start_time"] and timestamp <= mode["end_time"]:
+                if mode["status"] == Chassis.COMPLETE_AUTO_DRIVE:
+                    return True
+                else:
+                    return False
+        return False
+
+    def _bumps_rollback(self, bump_time):
+        """rollback 1 second when passing bumps"""
+
+        if bump_time - self._last_speed_down_2_time <= 1:
+            if self._check_status(self._last_speed_down_2_time):
+                self.auto_counts["speed_down_2"] -= 1
+            else:
+                self.manual_counts["speed_down_2"] -= 1
+
+        if bump_time - self._last_speed_up_2_time <= 1:
+            if self._check_status(self._last_speed_up_2_time):
+                self.auto_counts["speed_up_2"] -= 1
+            else:
+                self.manual_counts["speed_up_2"] -= 1
+
+        if bump_time - self._last_speed_down_4_time <= 1:
+            if self._check_status(self._last_speed_down_4_time):
+                self.auto_counts["speed_down_4"] -= 1
+            else:
+                self.manual_counts["speed_down_4"] -= 1
+
+        if bump_time - self._last_speed_up_4_time <= 1:
+            if self._check_status(self._last_speed_up_4_time):
+                self.auto_counts["speed_up_4"] -= 1
+            else:
+                self.manual_counts["speed_up_4"] -= 1
+
+        if bump_time - self._last_turning_time <= 1:
+            if self._check_status(self._last_turning_time):
+                self.auto_counts["turning"] -= 1
+            else:
+                self.manual_counts["turning"] -= 1
+
+
+    def calculate(self, bag_file):
+        with Bag(bag_file, 'r') as bag:
+            for topic, msg, _t in bag.read_messages([kLocalizationTopic]):
+                t = long(str(_t))*pow(10,-9)
+                #时间戳
+                #dt = time.strftime('%Y-%m-%d %H:%M:%S',time.localtime(_t))
+                self.timestamp = t
+                if t - self._bump_lock_time <= 3:
+                    continue
+                acc_x = msg.pose.linear_acceleration.x
+                acc_y = msg.pose.linear_acceleration.y
+                acc_z = msg.pose.linear_acceleration.z
+                diff_bump_time = t - self._last_bump_time
+                if abs(acc_z) >= 2 and diff_bump_time >= 3:
```

Comment:
make 2 and 3 const variable

Reply:
python doesn't have "const" like cpp. Have to define a new class if const variable is needed.

Reply:
similar to kMaxAccleration